taichi/program/sparse_solver.cpp

```diff
@@ -67,6 +67,176 @@ bool EigenSparseSolver<EigenSolver, EigenMatrix>::info() {
   return solver_.info() == Eigen::Success;
 }
 
+CuSparseSolver::CuSparseSolver() {
+#if defined(TI_WITH_CUDA)
+  if (!CUSPARSEDriver::get_instance().is_loaded()) {
+    bool load_success = CUSPARSEDriver::get_instance().load_cusparse();
+    if (!load_success) {
+      TI_ERROR("Failed to load cusparse library!");
+    }
+  }
+  if (!CUSOLVERDriver::get_instance().is_loaded()) {
+    bool load_success = CUSOLVERDriver::get_instance().load_cusolver();
+    if (!load_success) {
+      TI_ERROR("Failed to load cusolver library!");
+    }
+  }
+#endif
+}
+
+void CuSparseSolver::solve_cu(Program *prog,
+                              const SparseMatrix &sm,
+                              const Ndarray &b,
+                              Ndarray &x) {
+#ifdef TI_WITH_CUDA
+  cusparseHandle_t cusparseHandle = NULL;
+  CUSPARSEDriver::get_instance().cpCreate(&cusparseHandle);
+  cusolverSpHandle_t handle = NULL;
+  CUSOLVERDriver::get_instance().csSpCreate(&handle);
+
+  int major_version, minor_version, patch_level;
+  CUSOLVERDriver::get_instance().csGetProperty(MAJOR_VERSION, &major_version);
+  CUSOLVERDriver::get_instance().csGetProperty(MINOR_VERSION, &minor_version);
+  CUSOLVERDriver::get_instance().csGetProperty(PATCH_LEVEL, &patch_level);
+  printf("Cusolver version: %d.%d.%d\n", major_version, minor_version,
+         patch_level);
+
+  const cusparseSpMatDescr_t *A =
+      (const cusparseSpMatDescr_t *)(sm.get_matrix());
+  size_t nrows = 0, ncols = 0, nnz = 0;
+  void *drow_offsets = NULL, *dcol_indices = NULL, *dvalues = NULL;
+  cusparseIndexType_t csrRowOffsetsType, csrColIndType;
+  cusparseIndexBase_t idxBase;
+  cudaDataType valueType;
+  CUSPARSEDriver::get_instance().cpCsrGet(
+      *A, &nrows, &ncols, &nnz, &drow_offsets, &dcol_indices, &dvalues,
+      &csrRowOffsetsType, &csrColIndType, &idxBase, &valueType);
+
+  size_t db = prog->get_ndarray_data_ptr_as_int(&b);
+  size_t dx = prog->get_ndarray_data_ptr_as_int(&x);
+
+  float *h_z = (float *)malloc(sizeof(float) * ncols);
+  float *h_x = (float *)malloc(sizeof(float) * ncols);
+  float *h_b = (float *)malloc(sizeof(float) * nrows);
+  float *h_Qb = (float *)malloc(sizeof(float) * nrows);
+
+  int *h_Q = (int *)malloc(sizeof(int) * ncols);
+  int *hrow_offsets_B = (int *)malloc(sizeof(int) * (nrows + 1));
+  int *hcol_indices_B = (int *)malloc(sizeof(int) * nnz);
+  float *h_val_B = (float *)malloc(sizeof(float) * nnz);
+  int *h_mapBfromA = (int *)malloc(sizeof(int) * nnz);
```

Comment:
I found that `h_z`, `hrow_offsets_B `, `hcol_indices_B ` and `h_val_B ` are not deleted in the end of the function. Could you re-check it? :-)